### ISSUE TITLE
quad printing does not work as documented

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1610]
Migrated from [rt.perl.org#2846](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2846) (status was 'resolved')

Searchable as RT2846$

**Comment by p5pRT** [comment]
### From joseph@dixie.cscaper.com

#### Created by joseph@dixie.cscaper.com



from perlfunc&#8203;:
&nbsp;               If Perl understands "quads" \(64\-bit integers\)
&nbsp;               \(this requires either that the platform natively
&nbsp;               support quads or that Perl be specifically
&nbsp;               compiled to support quads\)\, the characters

&nbsp;                       d u o x X b i D U O

&nbsp;               print quads\, and they may optionally be preceded
&nbsp;               ^^^^^^^^^^^
&nbsp;               by

&nbsp;                       ll L q

&nbsp;               For example

&nbsp;                       %lld %16LX %qo

So\, what does "print quads" mean?  Here's two theories&#8203;:

Hypothesis 1&#8203;: Prints the correct value of quads\.
Hypothesis 2&#8203;: Prints the incorrect value of quads\.

On my machine it turns out to be Hypothesis 2&#8203;:

/usr/local/devel/perl\-longlong/bin/perl \-e '$foo = 5000000000; \\
&nbsp;                     printf "w/o&#8203;: %d  w&#8203;: %lld\\n"\, $foo\, $foo'
w/o&#8203;: 705032704  w&#8203;: 5000000000

&nbsp;  \-joseph

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by joseph at Thu Mar 23 23:02:23 MST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.12-20, archname=i586-linux-64all
    uname='linux dixie.cscaper.com 2.2.12-20 #1 mon sep 27 10:25:54 edt 1999 i586 unknown '
    config_args='-d'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=define use64bitall=define uselongdouble=define usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/devel/perl-longlong/lib/5.5.670/i586-linux-64all
    /usr/local/devel/perl-longlong/lib/5.5.670
    /usr/local/devel/perl-longlong/lib/site_perl/5.5.670/i586-linux-64all
    /usr/local/devel/perl-longlong/lib/site_perl/5.5.670
    /usr/local/devel/perl-longlong/lib/site_perl/5.5.670
    .


Environment for perl v5.6.0:
    HOME=/home/joseph
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=en_US
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/local/lib:/usr/X11R6/lib:/oracle/sw/lib
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/home/joseph/bin:/usr/local/bin:/usr/X11R6/bin:/usr/local/frame/bin:/usr/local/mysql/bin:/oracle/sw/bin:/usr/sbin:/usr/games:.
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @jhi


Change 5905 by jhi@&#8203;alpha on 2000/03/31 18&#8203;:50&#8203;:54

&nbsp;    Fix for "\[ID 20000329\.080\] quad printing does not work as documented"\,
&nbsp;    reported by Joseph N\. Hall\.  The bug was that even with use64bitint
&nbsp;    one had to use the prefixes q\, L\, or ll with printf/sprintf\,
&nbsp;    contrary to the documentation and common sense\.

Affected files \.\.\.

\.\.\. //depot/cfgperl/sv\.c\#102 edit
\.\.\. //depot/cfgperl/t/op/64bitint\.t\#2 edit

Differences \.\.\.

==== //depot/cfgperl/sv\.c\#102 \(text\) ====
Index&#8203;: perl/sv\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/sv.c.~1~	Fri Mar 31 21:56:32 2000
+++ perl/sv.c	Fri Mar 31 21:56:32 2000
@@ -5926,6 +5926,11 @@
 	    break;
 	}
 
+#ifdef USE_64_BIT_INT
+	if (!intsize)
+	    intsize = 'q';
+#endif
+
 	/* CONVERSION */
 
 	switch (c = *q++) {


```
</details>
==== //depot/cfgperl/t/op/64bitint.t#2 (text) ====
Index&#8203;: perl/t/op/64bitint\.t
<details><summary>Inline Patch</summary>

```diff
--- perl/t/op/64bitint.t.~1~	Fri Mar 31 21:56:32 2000
+++ perl/t/op/64bitint.t	Fri Mar 31 21:56:32 2000
@@ -10,13 +10,13 @@
 	unshift @INC, '../lib';
 }
 
-# This could use a lot of more tests.
+# This could use many more tests.
 
 # so that using > 0xfffffff constants and
 # 32+ bit integers don't cause noise
 no warnings qw(overflow portable);
 
-print "1..48\n";
+print "1..52\n";
 
 my $q = 12345678901;
 my $r = 23456789012;
@@ -239,4 +239,17 @@
 print "not " unless (0xf000000000000000 ^ 0xfffffffffffffff0) == 0x0ffffffffffffff0;
 print "ok 48\n";
 
+
+print "not " unless (sprintf "%b", ~0)   eq '1111111111111111111111111111111111111111111111111111111111111111';
+print "ok 49\n";
+
+print "not " unless (sprintf "%64b", ~0) eq '1111111111111111111111111111111111111111111111111111111111111111';
+print "ok 50\n";
+
+print "not " unless (sprintf "%d", ~0>>1) eq '9223372036854775807';
+print "ok 51\n";
+
+print "not " unless (sprintf "%u", ~0)    eq '18446744073709551615';
+print "ok 52\n";
+
 # eof
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From @jhi


In addition to the patch I already sent\, also this is required
in 32\-bit systems doing use64bitint\.

Change 5907 by jhi@&#8203;alpha on 2000/03/31 23&#8203;:42&#8203;:56

&nbsp;    pp\_gmtime\(\) assumed that newSVpvf %d would take C ints\,
&nbsp;    an assumption which \#5905 demolished\.

Affected files \.\.\.

\.\.\. //depot/cfgperl/pp\_sys\.c\#96 edit

Differences \.\.\.

==== //depot/cfgperl/pp\_sys\.c\#96 \(text\) ====
Index&#8203;: perl/pp\_sys\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/pp_sys.c.~1~	Sat Apr  1 02:43:05 2000
+++ perl/pp_sys.c	Sat Apr  1 02:43:05 2000
@@ -4086,14 +4086,14 @@
 	SV *tsv;
 	if (!tmbuf)
 	    RETPUSHUNDEF;
-	tsv = Perl_newSVpvf(aTHX_ "%s %s %2d %02d:%02d:%02d %d",
+	tsv = Perl_newSVpvf(aTHX_ "%s %s %2"IVdf" %02"IVdf":%02"IVdf":%02"IVdf" %"IVdf,
 			    dayname[tmbuf->tm_wday],
 			    monname[tmbuf->tm_mon],
-			    tmbuf->tm_mday,
-			    tmbuf->tm_hour,
-			    tmbuf->tm_min,
-			    tmbuf->tm_sec,
-			    tmbuf->tm_year + 1900);
+			    (IV)tmbuf->tm_mday,
+			    (IV)tmbuf->tm_hour,
+			    (IV)tmbuf->tm_min,
+			    (IV)tmbuf->tm_sec,
+			    (IV)tmbuf->tm_year + 1900);
 	PUSHs(sv_2mortal(tsv));
     }
     else if (tmbuf) {
End of Patch.
```
</details>
